### PR TITLE
Feat/fireworks ai

### PR DESCRIPTION
**Pull request by sk-portkey**
**Title:** 
- Integration of Fireworks AI (https://fireworks.ai/)

**Description:** (optional)
- Integrating Chat Complete, Embeddings, and Image generations

**Related Issues:** (optional)
- #251

### CODE REVIEW

File: src/providers/fireworks-ai/index.ts [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+} from './imageGenerate';
+
+const FireworksAIConfig: ProviderConfigs = {
+  complete: FireworksAICompleteConfig,
+  chatComplete: FireworksAIChatCompleteConfig,
+  embed: FireworksAIEmbedConfig,
+  imageGenerate: FireworksAIImageGenerateConfig,
+  api: FireworksAIAPIConfig,
+  responseTransforms: {
+    complete: FireworksAICompleteResponseTransform,
```
Comment (VisargD) [inline src/providers/fireworks-ai/index.ts:29]: stream-complete transformer function is present but it not mapped here. Can you please check once?

Comment (sk-portkey) [reply src/providers/fireworks-ai/index.ts:29]: done

File: src/providers/fireworks-ai/chatComplete.ts [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+  response: string
+) => string = (responseChunk) => {
+  let chunk = responseChunk.trim();
+  chunk = chunk.replace(/^data: /, '');
+  chunk = chunk.trim();
+  if (chunk === '[DONE]') {
+    return `data: ${chunk}\n\n`;
+  }
+  const parsedChunk: FireworksAIStreamChunk = JSON.parse(chunk);
+  return (
```
Comment (VisargD) [inline src/providers/fireworks-ai/chatComplete.ts:204]: As per their docs, fireworks provides token usage in the last stream chunk for streaming mode. Can we also attach usage in our last transformed stream chunk. We already do something similar for bedrock stream transformer.

Comment (sk-portkey) [reply src/providers/fireworks-ai/chatComplete.ts:204]: done

File: src/providers/fireworks-ai/imageGenerate.ts [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+import { ErrorResponse, ImageGenerateResponse, ProviderConfig } from '../types';
+import { generateInvalidProviderResponseError } from '../utils';
+import {
+  FireworksAIValidationErrorResponse,
+  FireworksAIErrorResponseTransform,
+  FireworksAIErrorResponse,
+} from './chatComplete';
+
+export const FireworksAIImageGenerateConfig: ProviderConfig = {
+  text_prompts: {
```
Comment (VisargD) [inline src/providers/fireworks-ai/imageGenerate.ts]: The OpenAI compatible param name is `prompt` which can then be mapped to `text_prompts`

File: src/providers/fireworks-ai/imageGenerate.ts [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+    param: 'text_prompts',
+    required: true,
+  },
+  model: {
+    param: 'model',
+    required: true,
+    default: 'stable-diffusion-xl-1024-v1-0',
+  },
+  height: {
```
Comment (VisargD) [inline src/providers/fireworks-ai/imageGenerate.ts]: `height` and `weight` are not directly available in OpenAI specs. You will have to use size and split it by `x`. Please check the OpenAI docs to compare the params

File: src/providers/fireworks-ai/imageGenerate.ts (hunk 3 above)
Comment (sk-portkey) [reply src/providers/fireworks-ai/imageGenerate.ts]: done

**Review by VisargD — DISMISSED** [review]

**Comment by vrushankportkey** [comment]
Also addresses #331

**Pull request merged**